from_dict/_from_dict.py

```diff
@@ -24,6 +25,39 @@ def __repr__(self):
         return f"RuntimeTypeError({self.location!r}, {self.expected_type!r}, {self.found_type!r})"
 
 
+@dataclass
+class Namespaces:
+    global_ns: Optional[dict]
+    local_ns: Optional[dict]
+    _types_cache: Optional[tuple] = None
+
+    def _get_types_in_namespaces(self):
```

Comment:
I will look into this but I don't think it going to work out well. Hashing all the elements in the namespace requires all the elements to be hashable. 
So the following example would not work because the dict in the locals is not hashable. 
```python
def example():
    # pretend that this is an import that is needed to convert the dict to OutputClass
    from json import JSONDecoder
    input_dict = {...}
    return from_dict(OutputClass, input_dict, globals(), locals())
```

Even if we found a way to get __hash__ to workaround objects that are not hashable, there is the issue that locals() will probably return a different dictionary most of the time.

Since ```typing.get_type_hints()``` just calls ```typing.get_type_hints``` I think it is safe to assume the current behavior will not change.  I can't think of a good way to guarantee that this will always be the case, so I see your concern. 

If you are worried about hash and equals not using all of information the namespaces. I would think it better to update ```Namespaces``` so that if either namespace is not None that it always has unique hash and is only equals to itself.  passing in local_ns or global_ns into from_dict() is needed so rarely it seems unnecessary to optimize for and the more logic we add for generating a hash the more it cuts into the performance benefits.

Reply:
Ah yeah, you raise a very good point.
Hmm..

Could we make Namespace's __init__ extract just the types from the namespaces and store just those? Then the contents of Namespace would be reflected in the hash 1:1 and there wouldn't be the risk that we talked about above.
I think this would solve everything, what do you think? (Maybe we'd want to rename Namespace to something like NamespaceTypes or something).

Reply:
That is a great idea. Thank you.